File: courses/codes/boring_business.py
well_plan = [[False for y in range(200)] for x in range(401)]


def drill(x, y, direction, steps):
    is_danger = False
    for i in range(steps):
        if direction == 'd':
            y -= 1
        elif direction == 'u':
            y += 1
        elif direction == 'l':
            x -= 1
        elif direction == 'r':
            x += 1
        else:
            print("Invalid direction")

        if well_plan[x][y]:
            is_danger = True
        else:
            well_plan[x][y] = True

    return x, y, is_danger


def print_well_plan(span_x, span_y):
    for j in range(-1, -span_y-1, -1):
        for i in range(-span_x-1, span_x+1):
            print('X' if well_plan[i][j] else ' ', end='')
        print()


# initial setup
current_x = 0
current_y = -1
current_danger = False
well_plan[current_x][current_y] = True

current_x, current_y, current_danger = drill(current_x, current_y, 'd', 2)
current_x, current_y, current_danger = drill(current_x, current_y, 'r', 3)
current_x, current_y, current_danger = drill(current_x, current_y, 'd', 2)
current_x, current_y, current_danger = drill(current_x, current_y, 'r', 2)
current_x, current_y, current_danger = drill(current_x, current_y, 'u', 2)
current_x, current_y, current_danger = drill(current_x, current_y, 'r', 2)
current_x, current_y, current_danger = drill(current_x, current_y, 'd', 4)
current_x, current_y, current_danger = drill(current_x, current_y, 'l', 8)
current_x, current_y, current_danger = drill(current_x, current_y, 'u', 2)

# print_well_plan(10, 10)
while True:
    cmd = input().split()
    if cmd[0] == 'q':
        break
    else:
        current_x, current_y, current_danger = drill(current_x, current_y, cmd[0], int(cmd[1]))
        # print_well_plan(10, 10)
        print(current_x, current_y, "DANGER" if current_danger else "safe")
        if current_danger:
            break

File: courses/codes/BoringBusiness.java
import java.util.*;

public class BoringBusiness {

    boolean[][] wellPlan = new boolean[401][200];
    int x = 200;    // -200 ... 200 is mapped to 0 ... 400
    int y = 199;    // -200 ... -1 is mapped to 0 ... 199
    boolean danger = false;

    public BoringBusiness() {
        // mark initial location
        wellPlan[this.x][this.y] = true;
    }

    public void drill(char direction, int steps) {
        for (int i = 0; i < steps; ++i) {
            if (direction == 'd') {
                this.y -= 1;
            } else if (direction == 'u') {
                this.y += 1;
            } else if (direction == 'l') {
                this.x -= 1;
            } else if (direction == 'r') {
                this.x += 1;
            } else {
                System.out.println("Invalid direction");
            }

            if (wellPlan[this.x][this.y]) {
                danger = true;
            } else {
                wellPlan[this.x][this.y] = true;
            }
        }
    }

    public void printWellPlan(int span_x, int span_y) {

        for (int j = 199; j > 199 - span_y; --j) {
            for (int i = 200 - span_x; i < 200 + span_x + 1; ++i) {
                System.out.print((wellPlan[i][j])? 'X': ' ');
            }
            System.out.println();
        }
    }

    public void printLocation() {
        System.out.print((this.x - 200) + ", " + (this.y - 200));
    }

    public static void main(String[] args) {

        BoringBusiness well = new BoringBusiness();

        // initial setup
        well.drill('d', 2);
        well.drill('r', 3);
        well.drill('d', 2);
        well.drill('r', 2);
        well.drill('u', 2);
        well.drill('r', 2);
        well.drill('d', 4);
        well.drill('l', 8);
        well.drill('u', 2);

        // well.printWellPlan(10, 10);

        Scanner scan = new Scanner(System.in);
        while (true) {
            // System.out.print("> ");
            String cmd = scan.next();
            int n = scan.nextInt();
            if (cmd.equals("q")) {
                break;
            } else {
                well.drill(cmd.charAt(0), n);
                // well.printWellPlan(10, 10);
                well.printLocation();
                System.out.println(" " + ((well.danger)? "DANGER": "safe"));
                if (well.danger) {
                    break;
                }
            }
        }
        scan.close();
    }
}
